### DISCUSSION TITLE
Building GINKGO on Windows - Any binaries ?

### DISCUSSION THREAD

**Discussion by shock-pro** [ginkgo-project/ginkgo#814] [Q&A]
Hello Everyone,
I read the Installation section and it is not clear to me whether
Ginkco builds on Windows with CMAKE ?
If yes, what toolset can be used, VS ?, MinGW?, etc.
Any prebuilt binaries available ?
Thanks in advance!

**Comment by shock-pro** [comment]
Sorry, I just found https://github.com/ginkgo-project/ginkgo/releases/tag/v1.3.0
where it clearly states that GINKGO can be built on Windows.

Still, I would like to know if there are any  prebuilt binaries available ?
Thanks

**Reply by upsj** [reply]
You can use MSVC cl, MinGW or Cygwin gcc. MSVC clang might also work, but we haven't tested that yet. The normal pattern
```bash
mkdir build
cd build
cmake ..
cmake . --build --config Release
```
should work, though you may want to add -j8 or something depending on the number of your cores to speed up the build.
While we do support Windows builds, we do not test them extensively. We only support 2 of our 5 backends (Reference and CUDA) with MSVC due to their outdated OpenMP support. 
If you want a quick start for working with Ginkgo on Windows, I would suggest using `vcpkg` or `conan`, which already have version 1.3.0 available, and `conan` might even provide prebuilt binaries? Though I personally prefer `vcpkg`, since their CMake integration is much nicer.
The thing with prebuilt binaries is that they require us to go all-or-nothing - Ginkgo has a number of optional features, and if we were to release binaries, they would need to all be enabled. That includes builds for all CUDA architectures (some of which may be deprecated or unavailable in newer or older CUDA versions, respectively), ROCm (which is unavailable on Windows), OpenMP (again, currently unavailable) and DPCPP (which should be available, but we haven't tested it yet)

**Comment by shock-pro** [comment]
Hello Tobias,
Thank you so much for your advice/recommendations.
You are absolutely right on MSVC's inferior OpenMP implementation.
Now that oneAPI is free, can we build on Windows with icl for all code
containing OpenMP, and leave cl just for nvcc (I don't know if you use
nvcc)?
Thanks again!

**Reply by upsj** [reply]
I think that might be difficult, as CMake does not support different C++ compilers at the same time. The only solution I could imagine would be using `cl` as `nvcc` host compiler and `icl` as regular C++ compiler. That does sound very complex though and almost like a recipe for binary incompatibility, but it might be worth a try? You can use MinGW gcc, which gives you OpenMP support, or use MSVC and get CUDA support, but both doesn't seem possible at the moment. Luckily, the situation is better under Linux (I cannot recommend WSL2 with GPU passthrough enough!)

**Comment by shock-pro** [comment]
I am using icl for all cpp code except .cu files, as nvcc only supports cl,
and the binaries are 100% compatible, that was always the idea behind icl.
However, I write my own Makefiles, so I can pick what compiler to use,
namely icl for all cpp, and nvcc for .cu, and nvcc automatically picks up
cl because that's all that it knows. I would have thought that cmake could
do the same when writing the VS projects, but based on your statement, it
can't.
Thanks again for your help.
Cheers,
